xla/hlo/transforms/collectives/collective_permute_combiner_test.cc

```diff
@@ -0,0 +1,348 @@
+    auto constant = b->AddInstruction(
+        HloInstruction::CreateConstant(LiteralUtil::CreateR0(42.3)));
+    Shape shape = ShapeUtil::MakeShape(
+        F32, {static_cast<int32_t>(sizes_in_kib.at(i) * 1024 / sizeof(float))});
+    auto input =
+        b->AddInstruction(HloInstruction::CreateBroadcast(shape, constant, {}));
+    inputs->push_back(input);
+    collective_permutes.push_back(
+        b->AddInstruction(HloInstruction::CreateCollectivePermute(
+            shape, input, source_target_pairs, /*channel_id=*/nullopt)));
+  }
+  return b->AddInstruction(HloInstruction::CreateTuple(collective_permutes));
+}
+
+using CollectivePermuteCombinerTest = HloHardwareIndependentTestBase;
+
+// Tests combination of several CollectivePermute instructions.
+TEST_F(CollectivePermuteCombinerTest, CombineCollectivePermutes) {
+  auto module = CreateNewVerifiedModule();
+
+  HloComputation::Builder b(TestName());
+  std::vector<HloInstruction*> inputs;
+  std::vector<std::pair<int64_t, int64_t>> source_target_pairs{
+      {0, 1}, {1, 2}, {2, 3}};
+  auto root = MakeCollectivePermutes({1, 2, 10, 7, 6}, &inputs,
+                                     source_target_pairs, &b);
+  auto computation = module->AddEntryComputation(b.Build());
+
+  // Run the CollectivePermute combiner optimization pass.
+  CollectivePermuteCombiner combine(10 * 1024 * 1024, kMaxCombineCount);
+  ASSERT_EQ(CollectivePermuteCount(*module), inputs.size());
+  TF_ASSERT_OK_AND_ASSIGN(bool changed, combine.Run(module.get()));
+  ASSERT_EQ(CollectivePermuteCount(*module), 1);
+  EXPECT_TRUE(changed);
+
+  ASSERT_EQ(root, computation->root_instruction());
+  ASSERT_EQ(inputs.size(), root->operands().size());
+
+  HloInstruction* combined = nullptr;
+  for (int64_t i = 0; i < root->operands().size(); ++i) {
+    HloInstruction* hlo = root->mutable_operand(i);
+    ASSERT_TRUE(hlo->opcode() == HloOpcode::kGetTupleElement);
+    EXPECT_EQ(hlo->tuple_index(), i);
+    EXPECT_TRUE(ShapeUtil::Equal(inputs[i]->shape(), hlo->shape()));
+
+    if (combined == nullptr) {
+      // Verify the combined collective permute instruction.
+      combined = hlo->mutable_operand(0);
+      ASSERT_TRUE(combined->opcode() == HloOpcode::kCollectivePermute);
+      EXPECT_TRUE(ShapeUtil::Equal(root->shape(), combined->shape()));
+      ASSERT_EQ(combined->operands().size(), inputs.size());
+    }
+    EXPECT_EQ(combined, hlo->operand(0));
+    EXPECT_TRUE(ShapeUtil::Equal(inputs[i]->shape(), hlo->shape()));
+    EXPECT_EQ(combined->operand(i), inputs[i]);
+    EXPECT_EQ(1, inputs[i]->users().size());
+  }
+  ASSERT_NE(combined, nullptr);
+}
+
+// Tests that the combination threshold is respected.
+TEST_F(CollectivePermuteCombinerTest, RespectThreshold) {
+  auto module = CreateNewVerifiedModule();
+
+  HloComputation::Builder b(TestName());
+  std::vector<HloInstruction*> inputs;
+  std::vector<std::pair<int64_t, int64_t>> source_target_pairs{
+      {0, 1}, {1, 2}, {2, 3}};
+  auto root =
```

Comment:
We got an unused variable error here. Could you please fix?
```suggestion
  [[maybe_unused]] auto root =
```

Reply:
Never mind. I'll make changes internally.